### PR TITLE
Add flag -swift-async-framepointer=auto,never,always to control swift extended frame information emission

### PR DESCRIPTION
**Pull request by aschwaighofer**
On linux we default to disable the extended frame info (since the system
libraries don't support it).

On darwin the default is to automatically choose based on the deployment target.

The Concurrency library explicitly forces extended frame information and the
back deployment library explicitly disables it.

### CODE REVIEW

**Comment by aschwaighofer** [comment]
Requires: https://github.com/apple/llvm-project/pull/3236

File: lib/Frontend/CompilerInvocation.cpp [hunk 1]
```diff
@@ -1896,6 +1896,24 @@ static bool ParseIRGenArgs(IRGenOptions &Opts, ArgList &Args,
     Opts.EnableGlobalISel = true;
   }
 
+  // Default to disabling swift async extended frame info on linux.
+  if (Triple.getOS() == llvm::Triple::Linux) {
```
Comment (DougGregor) [inline lib/Frontend/CompilerInvocation.cpp:1900]: Should this be `!Triple.getOS().isDarwin()`? I don't think the stack walkers anywhere else can cope with the extended frame info.

File: stdlib/public/Concurrency/CMakeLists.txt [hunk 2]
```diff
@@ -31,6 +31,25 @@ if(SWIFT_CONCURRENCY_USES_DISPATCH)
...
+set(SWIFT_RUNTIME_CONCURRENCY_C_FLAGS)
+set(SWIFT_RUNTIME_CONCURRENCY_SWIFT_FLAGS)
+
+if(NOT swift_concurrency_async_fp_mode)
+  set(swift_concurrency_async_fp_mode "always")
+endif()
+
+# Don't emit extended frame info on linux, system backtracer and system debugger
+# are unlikely to support it.
+if(CMAKE_SYSTEM_NAME STREQUAL Linux)
```
Comment (DougGregor) [inline stdlib/public/Concurrency/CMakeLists.txt:44]: Same comment here, it feels like Darwin is the special one here

**Review by DougGregor — APPROVED** [review]
Some minor comments, but this is looking great to me!

File: lib/Frontend/CompilerInvocation.cpp (hunk 1 above)
Comment (aschwaighofer) [reply lib/Frontend/CompilerInvocation.cpp:1900]: Good point

File: stdlib/public/Concurrency/CMakeLists.txt (hunk 2 above)
Comment (aschwaighofer) [reply stdlib/public/Concurrency/CMakeLists.txt:44]: Good point

**Comment by aschwaighofer** [comment]
Test with  apple/llvm-project#3236

@swift-ci Please test

**Comment by DougGregor** [comment]
Oh, I also think we'll need to thread this argument through the Clang importer, so the Clang instance gets the same setting

**Pull request closed without merging**